core/src/main/java/jenkins/util/TreeString.java

```diff
@@ -124,8 +124,8 @@ public boolean equals(final Object rhs) {
     public int hashCode() {
         int h = parent == null ? 0 : parent.hashCode();
 
-        for (int i = 0; i < label.length; i++) {
-            h = 31 * h + label[i];
+        for (char c : label) {
```

Comment:
Can replace this with `h = 31 * h + Arrays.hashCode(label)`

Reply:
Addressed in 72e2c6f below

Reply:
`h` was multiplied by 31 at every step. Now it's multiplied exactly once. Some elements of `parent` would be multiplied be the same power of 31 as the elements of the `label`, increasing chances of a collision (`a.bc` and `b.ac` would have the same hash `31 * 'a' + 31 * 'b' + 'c'`).

A good implementation of `hashCode` should take the seed as an argument rather than leave hash combining to the caller.